File: chat-client/src/main/java/com/tempchat/client/service/StompClient.java
package com.tempchat.client.service;

import com.fasterxml.jackson.databind.ObjectMapper;
import org.java_websocket.client.WebSocketClient;
import org.java_websocket.handshake.ServerHandshake;

import java.net.URI;
import java.util.Map;
import java.util.concurrent.atomic.AtomicInteger;
import java.util.function.BiConsumer;
import java.util.function.Consumer;

/**
 * Minimal STOMP-over-WebSocket client.
 * Connects to ws://host/ws/websocket (SockJS raw WebSocket endpoint).
 *
 * onMessage receives (sender, content) so the caller can decrypt content
 * before displaying it.
 */
public class StompClient extends WebSocketClient {

    private static final String STOMP_CONNECT =
            "CONNECT\naccept-version:1.2\nheart-beat:0,0\n\n\0";

    private final AtomicInteger subId = new AtomicInteger(0);
    private final ObjectMapper mapper = new ObjectMapper().findAndRegisterModules();

    private final BiConsumer<String, String> onMessage;
    private final Consumer<String> onUserEvent;
    private final String roomCode;
    private final String username;

    public StompClient(String wsUrl, String roomCode, String username,
                       BiConsumer<String, String> onMessage,
                       Consumer<String> onUserEvent) throws Exception {
        super(new URI(wsUrl));
        this.roomCode = roomCode;
        this.username = username;
        this.onMessage = onMessage;
        this.onUserEvent = onUserEvent;
    }

    @Override
    public void onOpen(ServerHandshake hs) {
        send(STOMP_CONNECT);
    }

    @Override
    public void onMessage(String frame) {
        if (frame.startsWith("CONNECTED")) {
            subscribeToTopic("/topic/chat/" + roomCode);
            subscribeToTopic("/topic/users/" + roomCode);
            sendStompMessage("/app/join/" + roomCode, "\"" + username + "\"");
            return;
        }
        if (frame.startsWith("MESSAGE")) {
            String body = extractBody(frame);
            String dest = extractHeader(frame, "destination");
            if (dest != null && dest.startsWith("/topic/users/")) {
                onUserEvent.accept(body.replace("\"", ""));
            } else {
                try {
                    var node = mapper.readTree(body);
                    String sender  = node.has("sender")  ? node.get("sender").asText()  : "?";
                    String content = node.has("content") ? node.get("content").asText() : body;
                    onMessage.accept(sender, content);
                } catch (Exception e) {
                    onMessage.accept("?", body);
                }
            }
        }
    }

    public void sendChatMessage(String content) {
        try {
            String json = mapper.writeValueAsString(Map.of(
                    "sender",   username,
                    "content",  content,
                    "roomCode", roomCode));
            sendStompMessage("/app/chat/" + roomCode, json);
        } catch (Exception e) {
            throw new RuntimeException(e);
        }
    }

    public void leaveRoom() {
        try {
            sendStompMessage("/app/leave/" + roomCode, "\"" + username + "\"");
        } catch (Exception ignored) {}
    }

    @Override public void onClose(int code, String reason, boolean remote) {}
    @Override public void onError(Exception ex) {}

    // ---- STOMP helpers ----

    private void subscribeToTopic(String topic) {
        String frame = "SUBSCRIBE\nid:sub-" + subId.getAndIncrement() +
                "\ndestination:" + topic + "\n\n\0";
        send(frame);
    }

    private void sendStompMessage(String destination, String body) {
        String frame = "SEND\ndestination:" + destination +
                "\ncontent-type:application/json\n\n" + body + "\0";
        send(frame);
    }

    private String extractBody(String frame) {
        int idx = frame.indexOf("\n\n");
        if (idx < 0) return frame;
        String raw = frame.substring(idx + 2);
        return raw.endsWith("\0") ? raw.substring(0, raw.length() - 1) : raw;
    }

    private String extractHeader(String frame, String headerName) {
        for (String line : frame.split("\n")) {
            if (line.startsWith(headerName + ":")) {
                return line.substring(headerName.length() + 1).trim();
            }
        }
        return null;
    }
}

File: cross-platform-client/Services/StompClient.cs
using System.Net.WebSockets;
using System.Text;
using System.Text.Json;

namespace TempChat.Services;

/// <summary>
/// Minimal STOMP-over-WebSocket client.
/// Connects to ws://host:port/ws/websocket (SockJS raw endpoint).
///
/// onMessage receives (sender, content) so the caller can decrypt before display.
/// </summary>
public sealed class StompClient : IDisposable
{
    private readonly ClientWebSocket _ws = new();
    private readonly Uri _wsUri;
    private readonly string _roomCode;
    private readonly string _username;
    private readonly Action<string, string> _onMessage;
    private readonly Action<string> _onUserEvent;
    private readonly CancellationTokenSource _cts = new();
    private int _subId;

    public StompClient(string wsUrl, string roomCode, string username,
        Action<string, string> onMessage, Action<string> onUserEvent)
    {
        _wsUri       = new Uri(wsUrl);
        _roomCode    = roomCode;
        _username    = username;
        _onMessage   = onMessage;
        _onUserEvent = onUserEvent;
    }

    public async Task ConnectAsync()
    {
        await _ws.ConnectAsync(_wsUri, _cts.Token);
        await SendRawAsync("CONNECT\naccept-version:1.2\nheart-beat:0,0\n\n\0");
        _ = ReadLoopAsync();
    }

    public async Task SendChatMessageAsync(string content)
    {
        string json = JsonSerializer.Serialize(new
        {
            sender   = _username,
            content,
            roomCode = _roomCode
        });
        await SendStompAsync($"/app/chat/{_roomCode}", json);
    }

    public async Task LeaveRoomAsync()
    {
        try { await SendStompAsync($"/app/leave/{_roomCode}", $"\"{_username}\""); }
        catch { /* best-effort */ }
    }

    // ── internals ─────────────────────────────────────────────────────

    private async Task SendStompAsync(string destination, string body)
    {
        string frame = $"SEND\ndestination:{destination}\ncontent-type:application/json\n\n{body}\0";
        await SendRawAsync(frame);
    }

    private async Task SendRawAsync(string frame)
    {
        byte[] bytes = Encoding.UTF8.GetBytes(frame);
        await _ws.SendAsync(bytes, WebSocketMessageType.Text, true, _cts.Token);
    }

    private async Task ReadLoopAsync()
    {
        var buffer = new byte[8192];
        var sb     = new StringBuilder();

        try
        {
            while (_ws.State == WebSocketState.Open && !_cts.IsCancellationRequested)
            {
                WebSocketReceiveResult result;
                do
                {
                    result = await _ws.ReceiveAsync(buffer, _cts.Token);
                    if (result.MessageType == WebSocketMessageType.Close) return;
                    sb.Append(Encoding.UTF8.GetString(buffer, 0, result.Count));
                }
                while (!result.EndOfMessage);

                foreach (string frame in sb.ToString().Split('\0'))
                    if (!string.IsNullOrWhiteSpace(frame))
                        ProcessFrame(frame.Trim('\r', '\n', ' '));

                sb.Clear();
            }
        }
        catch (OperationCanceledException) { }
        catch (Exception) { }
    }

    private void ProcessFrame(string frame)
    {
        if (frame.StartsWith("CONNECTED"))
        {
            _ = Task.Run(async () =>
            {
                await SubscribeAsync($"/topic/chat/{_roomCode}");
                await SubscribeAsync($"/topic/users/{_roomCode}");
                await SendStompAsync($"/app/join/{_roomCode}", $"\"{_username}\"");
            });
            return;
        }

        if (!frame.StartsWith("MESSAGE")) return;

        string body = ExtractBody(frame);
        string dest = ExtractHeader(frame, "destination") ?? "";

        if (dest.StartsWith("/topic/users/"))
        {
            _onUserEvent(body.Trim('"', '\r', '\n'));
            return;
        }

        try
        {
            using var doc  = JsonDocument.Parse(body);
            var root       = doc.RootElement;
            string sender  = root.TryGetProperty("sender",  out var s) ? s.GetString() ?? "?" : "?";
            string content = root.TryGetProperty("content", out var c) ? c.GetString() ?? body : body;
            _onMessage(sender, content);
        }
        catch
        {
            _onMessage("?", body);
        }
    }

    private async Task SubscribeAsync(string topic)
    {
        int id = Interlocked.Increment(ref _subId);
        await SendRawAsync($"SUBSCRIBE\nid:sub-{id}\ndestination:{topic}\n\n\0");
    }

    private static string ExtractBody(string frame)
    {
        int idx = frame.IndexOf("\n\n", StringComparison.Ordinal);
        return idx < 0 ? frame : frame[(idx + 2)..];
    }

    private static string? ExtractHeader(string frame, string name)
    {
        foreach (string line in frame.Split('\n'))
            if (line.StartsWith(name + ":"))
                return line[(name.Length + 1)..].Trim();
        return null;
    }

    public void Dispose()
    {
        _cts.Cancel();
        _ws.Dispose();
        _cts.Dispose();
    }
}
